README.md
# regtabletotext: helpers to print regression output as text

## What is regtabletotext?

This package is a collection of helper functions to print regression output of the Python packages `statsmodels`, `linearmodels`, and `arch` as text strings. The helpers are particularly useful for users who want to render regression output in [Quarto](https://quarto.org/) to HTML and PDF.

If you want to export your results to LaTeX or HTML, please check out [stargazer](https://pypi.org/project/stargazer/).

## How to install regtabletotext?

The package is available on [pypi.org/project/regtabletotext](https://pypi.org/project/regtabletotext/):

```
pip install regtabletotext
```

## How to use regtabletotext?

Currently supported model types:

- `statsmodels.regression.linear_model.RegressionResultsWrapper`
- `linearmodels.panel.results.PanelEffectsResults`
- `arch.univariate.base.ARCHModelResult`

### For statsmodels regression output

The following code chunk
```
import pandas as pd
import numpy as np
import statsmodels.formula.api as smf
from regtabletotext import prettify_result

n = 100 
data = pd.DataFrame({
    'age': np.random.randint(18, 65, n),
    'income': np.random.normal(50, 20, n) ,
    'education': np.random.randint(12, 22, n),
    'hours_worked': np.random.randint(20, 60, n),
    'satisfaction': np.random.randint(1, 11, n) 
})

mod = smf.ols(formula='income ~ age + education + hours_worked', data=data)
res = mod.fit()

prettify_result(res)
```

returns this text
```
OLS Model:
Lottery ~ Literacy + Wealth + Region

Coefficients:
             Estimate  Std. Error  Statistic  p-Value
Intercept      38.652       9.456      4.087    0.000
Region[T.E]   -15.428       9.727     -1.586    0.117
Region[T.N]   -10.017       9.260     -1.082    0.283
Region[T.S]    -4.548       7.279     -0.625    0.534
Region[T.W]   -10.091       7.196     -1.402    0.165
Literacy       -0.186       0.210     -0.886    0.378
Wealth          0.452       0.103      4.390    0.000

Summary statistics:
- Number of observations: 85
- R-squared: 0.338, Adjusted R-squared: 0.287
- F-statistic: 6.636 on 6 and 78 DF, p-value: 0.000
```

### For linearmodels regression output

This code chunk
```
import pandas as pd
import numpy as np
from linearmodels import PanelOLS
from regtabletotext import prettify_result

np.random.seed(42)
n_entities = 100
n_time_periods = 5

data = {
    'entity': np.repeat(np.arange(n_entities), n_time_periods),
    'time': np.tile(np.arange(n_time_periods), n_entities),
    'y': np.random.randn(n_entities * n_time_periods),
    'x1': np.random.randn(n_entities * n_time_periods),
    'x2': np.random.randn(n_entities * n_time_periods)
}

df = pd.DataFrame(data)
df.set_index(['entity', 'time'], inplace=True)

formula = 'y ~ x1 + x2 + EntityEffects + TimeEffects'
model = PanelOLS.from_formula(formula, df)
result = model.fit()

prettify_result(result, options={'digits':2, 'include_residuals': True})
```
produces this output
```
Panel OLS Model:
y ~ x1 + x2 + EntityEffects + TimeEffects

Covariance Type: Unadjusted

Residuals:
 Mean  Std  Min   25%  50%  75%  Max
  0.0 0.87 -2.7 -0.62  0.0 0.58 3.16

Coefficients:
    Estimate  Std. Error  t-Statistic  p-Value

x1     -0.06        0.05        -1.12     0.26
x2     -0.04        0.05        -0.80     0.42

Included Fixed Effects:
        Total
Entity  100.0
Time      5.0

Summary statistics:
- Number of observations: 500
- R-squared (incl. FE): 0.21, Within R-squared: 0.01
- F-statistic: 1.05, p-value: 0.35
```
For multiple models, you can also use `prettify_result`:
```
formula = 'y ~ x1 + x2'
model = PanelOLS.from_formula(formula, df)
result1 = model.fit()

formula = 'y ~ x2 + EntityEffects'
model = PanelOLS.from_formula(formula, df)
result2 = model.fit()

formula = 'y ~ x1 + x2 + EntityEffects + TimeEffects'
model = PanelOLS.from_formula(formula, df)
result3 = model.fit()

results = [result1, result2, result3]

prettify_result(results)
```
The result is:
```
Dependent var.        y               y               y

x1              -0.072 (-1.59)                  -0.058 (-1.12)
x2              -0.049 (-1.14)  -0.049 (-0.98)  -0.041 (-0.8)

Fixed effects                       Entity       Entity, Time
VCOV type         Unadjusted      Unadjusted      Unadjusted
Observations         500             500             500
R2 (incl. FE)       0.008           0.198           0.208
Within R2           0.006           0.002           0.006
```

### For arch estimation output

For arch estimations like these:
```
import datetime as dt
import pandas_datareader.data as web
from arch import arch_model
import arch.data.sp500
from regtabletotext import prettify_result

start = dt.datetime(1988, 1, 1)
end = dt.datetime(2018, 1, 1)
data = arch.data.sp500.load()
returns = 100 * data['Adj Close'].pct_change().dropna()
am_fit = arch_model(returns).fit(update_freq=5)

prettify_result(am_fit)
```
you get:
```
Constant Mean - GARCH Model Results

Mean Coefficients:
    Estimate  Std. Error  Statistic  p-Value
mu     0.056       0.011      4.906      0.0

Coefficients for GARCH(p: 1, q: 1):
          Estimate  Std. Error  Statistic  p-Value
omega        0.018       0.005      3.738      0.0
alpha[1]     0.102       0.013      7.852      0.0
beta[1]      0.885       0.014     64.125      0.0

Summary statistics:
- Number of observations: 5,030
- Distribution: Normal distribution
- Multiple R-squared: 0.000, Adjusted R-squared: 0.000
- BIC: 13,907.530,  AIC: 13,881.437
```

HISTORY.md
# Version 0.0.12
- Added note that table contains t-statistics
- `prettify_result()` now passes lists to `prettify_results()`

# Version 0.0.11
- Replaced reporting of F-statistic nan with "F-statistic not available"

# Version 0.0.10
- Increased default `max_width` from 64 to 80

# Version 0.0.9
- Added support for list of `linearmodels.panel.result.PanelEffectsresult` results 

# Version 0.0.8
- Added support for 'arch.univariate.base.ARCHModelResult'
- Simplified and extended `calculate_residuals_statistics()`
- Added covariance type and fixed effects table to `linearmodels.panel.result.PanelEffectsresult` print result
- Moved default values to function definitions
- Updated required install to pandas only
- Added support for `arch.univariate.base.ARCHModelResult`

# Version 0.0.7
- Moved type checks to functions and globals
- Removed unnecessary whitespace from model formula
- Introduced `max_width` with default to 64 characters 

# Version 0.0.6
- Introduced check if 't' column is present, otherwise use 'z' column to handle models with only one coefficient

# Version 0.0.5
- Replaced explicit options by flexible `**options` parameter
- Supported classes: `statsmodels.regression.linear_model.RegressionResultsWrapper` and `linearmodels.panel.result.PanelEffectsresult`
- Introduced `ALLOWED_OPTIONS` global

# Version 0.0.4
- First working version with statsmodels support

regtabletotext/sm.py
import pandas as pd

def calculate_residuals_statistics(results, **options):
    """
    Calculate and return the statistics of residuals from the given regression results.
    
    Parameters:
    - results (object): A regression results object that contains a 'resid' attribute, typically from statsmodels.
    - options (dict): Optional parameters for the function. Current options include:
        * 'digits': Number of decimal places to round the calculated statistics to.
    
    Returns:
    - pd.DataFrame: A DataFrame containing the following statistics of residuals:
        * Min: Minimum value of residuals.
        * Q25: 25th percentile (1st quartile) of residuals.
        * Q50: Median (50th percentile) of residuals.
        * Q75: 75th percentile (3rd quartile) of residuals.
        * Max: Maximum value of residuals.
    """
    
    # Check if the results object has the 'resid' attribute and is from statsmodels
    if not hasattr(results, 'resid') or not 'statsmodels' in str(type(results)):
        raise ValueError("The 'results' parameter should be a single regression results object from statsmodels with a 'resid' attribute.")
    
    # Extract options or set defaults
    digits = options.get('digits', 3)
    
    # Extract residuals
    residuals = results.resid
    stats = {
        "Min": residuals.min(),
        "Q25": residuals.quantile(0.25),
        "Q50": residuals.median(),
        "Q75": residuals.quantile(0.75),
        "Max": residuals.max()
    }

    residuals_stats = pd.DataFrame(stats, index=[0])
    
    # Apply options
    if 'digits' in options:
        residuals_stats = residuals_stats.round(digits)
    
    return residuals_stats

def create_coefficients_table(results, **options):
    """
    Extract and format the coefficients table from regression results.
    
    This function extracts the coefficients table from the summary of regression results, 
    typically from statsmodels, and returns a formatted DataFrame with selected columns 
    and rounded values.
    
    Parameters:
    ----------
    results : object
        A regression results object that contains a 'summary' method, typically from statsmodels.
    
    options : dict
        Optional parameters for the function. Current options include:
        * 'digits': Number of decimal places to round the values in the coefficients table to (default=3).

    Returns:
    -------
    pd.DataFrame: 
        A DataFrame containing the following columns:
        * Estimate: Coefficient estimates.
        * Std. Error: Standard error of the coefficients.
        * t-Statistic: t-statistic values.
        * p-Value: p-values associated with the t-statistics.
    """

     # Check if the results object is from statsmodels
    if not 'statsmodels' in str(type(results)):
        raise ValueError("The 'results' parameter should be a single regression results object from statsmodels.")
    
    # Extract options or set defaults
    digits = options.get('digits', 3)
    
    # Extract results 
    results_data = results.summary().tables[1].data

    # Collect coefficient statistics in a data frame
    coefficients_table = (pd.DataFrame(
        results_data[1:], 
        columns=results_data[0]
      )
      .get(["", "coef", "std err", "t", "P>|t|"])
      .rename(columns={
        "coef": "Estimate",
        "std err": "Std. Error",
        "t": "t-Statistic",
        "P>|t|": "p-Value"
        },
      )
      .set_index("")
      .apply(pd.to_numeric, errors='coerce')
      .round(digits)
    )
    
    return(coefficients_table)
  
def prettify_statsmodels(results, **options):
    """
    Format and print regression results in a style similar to R's summary() output for linear models.
    
    This function takes regression results from statsmodels and prints a summary 
    that resembles the output provided by R's summary() function for linear models. The summary 
    includes the model formula, coefficients table, and other summary statistics. Optionally, 
    it can also include residuals statistics.

    Parameters:
    ----------
    results : object
        A regression results object from statsmodels that contains attributes 
        like 'model', 'nobs', 'mse_resid', 'rsquared', 'rsquared_adj', 'fvalue', 'df_model', 
        'df_resid', and 'f_pvalue'.

    options : dict
        Optional parameters for the function. Current options include:
        * 'digits': Number of decimal places to round the values in the summary to (default=3).
        * 'include_residuals': Whether to include residuals in the output (default=False).

    Returns:
    -------
    None
        The function prints the formatted summary directly.
    """
    # Check if the results object is from statsmodels
    if not 'statsmodels' in str(type(results)):
        raise ValueError("The 'results' parameter should be a single regression results object from statsmodels.")
    
    # Extract options or set defaults
    digits = options.get('digits', 3)
    include_residuals = options.get('include_residuals', False)
    
    # Initialize the output string
    output = f"\nModel:\n{results.model.formula}\n\n"
    
    # Add residuals to the output string if required
    if include_residuals:
        output += f"Residuals:\n{calculate_residuals_statistics(results, digits=digits).to_string(index=False)}\n\n"

    # Add coefficients to the output string
    # Assuming you have a function called 'create_coefficients_table' that creates the coefficients table
    output += f"Coefficients:\n{create_coefficients_table(results, digits=digits).to_string()}\n\n"

    # Add footer with additional statistics to the output string
    output += (
        f"Summary statistics:\n"
        f"- Number of observations: {results.nobs:.0f}\n"
        f"- Residual standard error: {results.mse_resid**0.5:.{digits}f}\n"
        f"- Multiple R-squared: {results.rsquared:.{digits}f}, Adjusted R-squared: {results.rsquared_adj:.{digits}f}\n"
        f"- F-statistic: {results.fvalue:.{digits}f} on {results.df_model:.0f} and {results.df_resid:.0f} DF, p-value: {results.f_pvalue:.{digits}f}\n"
    )

    # Print the output string
    print(output)



regtabletotext/funs.py
# ====================
# SECTION: Imports
# ====================
import pandas as pd
import numpy as np
from tabulate import tabulate

# ====================
# SECTION: Constants
# ====================
ALLOWED_OPTIONS = {'digits', 'include_residuals', 'max_width'}
TYPE_STATSMODELS = 'statsmodels.regression.linear_model.RegressionResultsWrapper'
TYPE_LINEARMODELS = 'linearmodels.panel.results.PanelEffectsResults'
TYPE_ARCH_MODEL = 'arch.univariate.base.ARCHModelResult'
SUPPORTED_MODELS = {TYPE_STATSMODELS, TYPE_LINEARMODELS, TYPE_ARCH_MODEL}

# ===================
# SECTION: Validation Functions
# ====================
def is_result_type_valid(result):
    result_type = type(result).__module__ + "." + type(result).__name__
    return(result_type in SUPPORTED_MODELS)

def is_result_type_statsmodels(result):
    result_type = type(result).__module__ + "." + type(result).__name__
    return(result_type in TYPE_STATSMODELS)

def is_result_type_linearmodels(result):
    result_type = type(result).__module__ + "." + type(result).__name__
    return(result_type in TYPE_LINEARMODELS)

def are_result_type_linearmodels(results):
    return all(is_result_type_linearmodels(result) for result in results)

def is_result_type_arch_model(result):
    result_type = type(result).__module__ + "." + type(result).__name__
    return(result_type in TYPE_ARCH_MODEL)

# ===================
# SECTION: Helper Functions
# ====================
def clean_model_formula(model_formula, options={'max_width': 80}):
    """
    Cleans and formats a given model formula to fit within a specified width.

    This function takes a model formula as input, removes any extra spaces, and ensures that the 
    formula does not exceed the specified maximum width (`max_width`). If the formula exceeds the 
    `max_width`, it splits the formula at the last '+' sign before the `max_width` and continues 
    the formula on the next line.

    Parameters:
    ----------
    model_formula : str
        The model formula to be cleaned and formatted. Typically, this is a string representation 
        of a regression model formula, e.g., "y ~ x1 + x2 + x3".
    options (dict): Optional parameters for the function. Current options include:
        * 'max_width': The maximum width (number of characters) that the formula should occupy on a single line (default: 80). 

    Returns:
    -------
    str
        The cleaned and formatted model formula.
    """

    max_width = options.get('max_width')
        
    model_formula_cleaned = ' '.join(model_formula.split())

    if len(model_formula_cleaned) <= max_width:
        return model_formula_cleaned
    else:
        # Split the formula at the last '+' before max_width
        split_index = model_formula_cleaned.rfind('+', 0, max_width)
        return model_formula_cleaned[:split_index] + '\n + ' + model_formula_cleaned[split_index+1:].strip()

def calculate_residuals_statistics(residuals, options={'digits': 3}):
    """
    Calculate and return the statistics of residuals from the given regression result.
    
    Parameters:
    - residuals series, Pandas series type
    - options (dict): Optional parameters for the function. Current options include:
        * 'digits': Number of decimal places to round the calculated statistics to (default: 3).

    Returns:
    - pd.DataFrame: A DataFrame containing the following statistics of residuals:
        * Mean: Mean value of residuals.
        * Std: Standard deviation of residuals.
        * Min: Minimum value of residuals.
        * 25%: 25th percentile (1st quartile) of residuals.
        * 50%: Median (50th percentile) of residuals.
        * 75%: 75th percentile (3rd quartile) of residuals.
        * Max: Maximum value of residuals.
    """
   
    # Extract options or use defaults
    digits = options.get('digits')

    residuals_stats = residuals.describe().iloc[1:]
    residuals_stats.index = [i.capitalize() for i in residuals_stats.index]
    residuals_stats = pd.DataFrame(residuals_stats).T.round(digits)
    residuals_stats.index.name = None

    return residuals_stats

def truncate_coefficients_table(coefficients_table, options={'max_width': 80}):
    """
    Truncate and return coefficient table with max width.
    Parameters:
    - coefficients_table, Pandas DataFrame type
    - max_width, int number of string characters per line
    Returns:
    - pd.DataFrame: A DataFrame containing coefficients
    """
    
    max_width = options.get('max_width')

    # Truncate coefficient names if they exceed max_width
    for index, row in coefficients_table.iterrows():
        coeff_name = index
        values_str = ' '.join(map(str, row.values))

        # Check if the combined width exceeds max_width
        if len(coeff_name) + len(values_str) + 20 > max_width:
            # Calculate how much to truncate the coefficient name
            trunc_length = max_width - len(values_str) - 23
            truncated_name = coeff_name[:trunc_length] + "..."

            # Update the index in the DataFrame
            coefficients_table = coefficients_table.rename(
                index={coeff_name: truncated_name})
    return coefficients_table

def create_coefficients_table(result, options={'digits': 3, 'max_width': 80}):
    """
    Extract and format the coefficients table from regression result.
    
    This function extracts the coefficients table from the summary of regression result, 
    typically from statsmodels, and returns a formatted DataFrame with selected columns 
    and rounded values.
    
    Parameters:
    ----------
    result : object
        A regression result object that contains a 'summary' method, typically from statsmodels.
    
    options : dict
        Optional parameters for the function. Current options include:
        * 'digits': Number of decimal places to round the values in the coefficients table to (default=3).
        * 'max_width': The maximum width (number of characters) that the formula should occupy on a single line (default: 80). 

    Returns:
    -------
    list of
    pd.DataFrame: 
        A DataFrame containing the following columns:
        * Estimate: Coefficient estimates.
        * Std. Error: Standard error of the coefficients.
        * t-Statistic: t-statistic values.
        * p-Value: p-values associated with the t-statistics.
    """

    # Check if the result object is valid
    if not is_result_type_valid(result):
        raise ValueError("The 'result' parameter should be a single regression result object from statsmodels or linearmodels.")
    
    # Extract options or use defaults
    digits = options.get('digits')
    max_width = options.get('max_width')

    # Initialize output list
    coefficients_tables_list = []
    
    if is_result_type_statsmodels(result):
        # Extract result 
        result_data = result.summary().tables[1].data

        # Check if 't' column is present, otherwise use 'z' column
        column_to_use = 't' if 't' in result_data[0] else 'z'
        p_value_column = 'P>|t|' if 't' in result_data[0] else 'P>|z|'

        # Collect coefficient statistics in a data frame
        coefficients_table = (
            pd.DataFrame(result_data[1:], columns=result_data[0])
            .get(["", "coef", "std err", column_to_use, p_value_column])
            .rename(columns={
                "coef": "Estimate",
                "std err": "Std. Error",
                column_to_use: "t-Statistic",
                p_value_column: "p-Value"
            })
            .set_index("")
            .apply(pd.to_numeric, errors='coerce')
            .round(digits)
        )

    if is_result_type_linearmodels(result):
        # Extract result 
        result_data = result.summary.tables[1].data

        # Collect coefficient statistics in a data frame
        coefficients_table = (pd.DataFrame(
                result_data[1:], 
                columns=result_data[0]
            )
            .get(["", "Parameter", "Std. Err.", "T-stat", "P-value"])
            .rename(columns={
                "Parameter": "Estimate",
                "Std. Err.": "Std. Error",
                "T-stat": "t-Statistic",
                "P-value": "p-Value"
                },
            )
            .set_index("")
            .apply(pd.to_numeric, errors='coerce')
            .round(digits)
        )

    if is_result_type_arch_model(result):
        # Extract result
        result_data = result.summary().tables[1].data

        # Check if 't' column is present, otherwise use 'z' column
        column_to_use = 't' if 't' in result_data[0] else 'z'
        p_value_column = 'P>|t|' if 't' in result_data[0] else 'P>|z|'

        # Collect coefficient statistics in a data frame
        coefficients_table = (
            pd.DataFrame(result_data[1:], columns=result_data[0])
            .get(["", "coef", "std err", column_to_use, p_value_column])
            .rename(columns={
                "coef": "Estimate",
                "std err": "Std. Error",
                column_to_use: "t-Statistic",
                p_value_column: "p-Value"
            })
            .set_index("")
            .apply(pd.to_numeric, errors='coerce')
            .round(digits)
        )

        # Extract result
        result_data_vola = result.summary().tables[2].data

        # Collect coefficient statistics in a data frame
        coefficients_table_vola = (
            pd.DataFrame(result_data_vola[1:], columns=result_data_vola[0])
            .get(["", "coef", "std err", column_to_use, p_value_column])
            .rename(columns={
                "coef": "Estimate",
                "std err": "Std. Error",
                column_to_use: "t-Statistic",
                p_value_column: "p-Value"
            })
            .set_index("")
            .apply(pd.to_numeric, errors='coerce')
            .round(digits)
        )
        coefficients_table_vola = truncate_coefficients_table(coefficients_table_vola, options={'max_width': max_width})
        coefficients_table_vola.index.name = None

    coefficients_table = truncate_coefficients_table(coefficients_table, options={'max_width': max_width})
    coefficients_table.index.name = None
    coefficients_tables_list.append(coefficients_table)

    if is_result_type_arch_model(result):
        coefficients_tables_list.append(coefficients_table_vola)

    return coefficients_tables_list

def create_fixed_effects_table(result):
    """
    Create a table summarizing the fixed effects included in a panel model result.

    The table will show the total number of entities and time periods, but only for 
    the effects that are actually included in the model (based on result.included_effects).

    Parameters:
    - result (PanelEffectsResults): The results object obtained after fitting a panel data model using `linearmodels`.

    Returns:
    - DataFrame: A pandas DataFrame with the effect types (Entity, Time) as the index and the total counts as values.
                 Only the effects present in result.included_effects will be included in the DataFrame.
    """
    fixed_effects_table = pd.DataFrame({
        '': ['Entity', 'Time'],
        'Total': [result.entity_info.total, result.time_info.total]
    }).set_index('')
    fixed_effects_table = fixed_effects_table[fixed_effects_table.index.isin(result.included_effects)]
    fixed_effects_table['Total'] = fixed_effects_table['Total'].astype(int)
    fixed_effects_table.index.name = None

    return(fixed_effects_table)

# ===================
# SECTION: Main Functions
# ====================
def prettify_result(result, options={'digits': 3, 'include_residuals': False, 'max_width': 80}):
    """
    Format and print regression result in a style similar to R's summary() output for linear models.
    
    This function takes regression result from statsmodels and prints a summary 
    that resembles the output provided by R's summary() function for linear models. The summary 
    includes the model formula, coefficients table, and other summary statistics. Optionally, 
    it can also include residuals statistics.

    Parameters:
    ----------
    result : object
        A regression result object from statsmodels that contains attributes 
        like 'model', 'nobs', 'mse_resid', 'rsquared', 'rsquared_adj', 'fvalue', 'df_model', 
        'df_resid', and 'f_pvalue'.

    options : dict
        Optional parameters for the function. Current options include:
        * 'digits': Number of decimal places to round the values in the coefficients table to (default=3).
        * 'include_residuals': Whether to include residuals in the output (default=False).
        * 'max_width': The maximum width (number of characters) that the formula should occupy on a single line (default: 80). 

    Returns:
    -------
    None
        The function prints the formatted summary directly.
    """
    # Check if the result object is valid
    if not is_result_type_valid(result):
        raise ValueError("The 'result' parameter is currently not supported.")
    
    # Check if options are valid
    invalid_options = set(options.keys()) - ALLOWED_OPTIONS
    if invalid_options:
        raise ValueError(f"Invalid options provided: {', '.join(invalid_options)}")
    
    # Extract options or use defaults
    digits = options.get('digits', 3)
    include_residuals = options.get('include_residuals', False)
    max_width = options.get('max_width', 80)

    if is_result_type_statsmodels(result):
        # Initialize the output string
        if (hasattr(result.model, "formula")):
            model_formula = clean_model_formula(result.model.formula, options={'max_width': max_width})
            output = f"OLS Model:\n{model_formula}\n\n"
        else:
            output = f"OLS Model (no formula provided)\n\n"
        
        # Add residuals to the output string if required
        if include_residuals:
            residuals_statistics = calculate_residuals_statistics(result.resid, options={'digits': digits}).to_string(index=False)
            output += f"Residuals:\n{residuals_statistics}\n\n"

        # Add coefficients to the output string
        coefficients_table = create_coefficients_table(result, options={'digits': digits, 'max_width': max_width})[0].to_string()
        output += f"Coefficients:\n{coefficients_table}\n\n"

        # Add footer with additional statistics to the output string
        output += (
            f"Summary statistics:\n"
            f"- Number of observations: {result.nobs:,.0f}\n"
            f"- R-squared: {result.rsquared:.{digits}f}, Adjusted R-squared: {result.rsquared_adj:.{digits}f}\n"
        )
        if (np.isnan(result.fvalue)):
            output += f"- F-statistic not available\n"
        else:
            output += f"- F-statistic: {result.fvalue:,.{digits}f} on {result.df_model:.0f} and {result.df_resid:.0f} DF, p-value: {result.f_pvalue:.{digits}f}\n"

    
    if is_result_type_linearmodels(result):
        # Initialize the output string
        output = f"Panel OLS Model:\n{clean_model_formula(result.model.formula)}\n\n"
        
        # Add covariance type
        output += f"Covariance Type: {result._cov_type}\n\n"

        # Add residuals to the output string if required
        if include_residuals:
            # Assuming you have a function called 'calculate_residuals_statistics' for linearmodels
            residual_statistics = calculate_residuals_statistics(result.resids, options={'digits': digits}).to_string(index=False)
            output += f"Residuals:\n{residual_statistics}\n\n"

        # Add coefficients to the output string
        coefficients_table = create_coefficients_table(result, options={'digits': digits, 'max_width': max_width})[0].to_string()
        output += f"Coefficients:\n{coefficients_table}\n\n"

        # Include table with included fixed effects (if any)
        if len(result.included_effects) > 0:
            fixed_effects_table = create_fixed_effects_table(result).to_string()
            output += f"Included Fixed Effects:\n{fixed_effects_table}\n\n"            

        # Add footer with additional statistics to the output string
        output += (
            f"Summary statistics:\n"
            f"- Number of observations: {result.nobs:,.0f}\n"
            f"- R-squared (incl. FE): {result.rsquared_inclusive:.{digits}f}, Within R-squared: {result.rsquared_within:.{digits}f}\n"
        )
        if (np.isnan(result.f_statistic.stat)):
            f"- F-statistic not available\n"
        else:
            f"- F-statistic: {result.f_statistic.stat:,.{digits}f}, p-value: {result.f_statistic.pval:.{digits}f}\n"

    if is_result_type_arch_model(result):
        model_name = result.summary().as_text().split('\n')[0].strip()
        output = f"\n{model_name}\n\n"
        if include_residuals:
            residuals_statistics = calculate_residuals_statistics(result.resid, options={'digits': digits}).to_string(index=False)
            output += f"Residuals:\n{residuals_statistics}\n\n"

        # Add coefficients to the output string
        output += f"Mean Coefficients:\n{create_coefficients_table(result, options={'digits': digits, 'max_width': max_width})[0].to_string()}\n\n"
        output += f"Coefficients for {str(result.model.volatility)}:\n{create_coefficients_table(result, options={'digits': digits, 'max_width': max_width})[1].to_string()}\n\n"

        # Add footer with additional statistics to the output string
        output += (
            f"Summary statistics:\n"
            f"- Number of observations: {result.nobs:,.0f}\n"
            f"- Distribution: {str(result.model.distribution)}\n"
            f"- Multiple R-squared: {result.rsquared:.{digits}f}, Adjusted R-squared: {result.rsquared_adj:.{digits}f}\n"
            f"- BIC: {result.bic:,.{digits}f},  AIC: {result.aic:,.{digits}f}\n"
        )

    # Print the output string
    print(output)

def prettify_results(results, options={'digits': 3}):
    """
    Formats and prints a summary table of regression results from a list of 'linearmodels.panel.results.PanelEffectsResults' objects.
    
    The function prints a table that includes the dependent variables, estimated coefficients with t-statistics, fixed effects, 
    variance-covariance (VCOV) type, number of observations, inclusive R-squared, and within R-squared. Coefficients and t-statistics 
    are rounded to the specified number of digits. 
    
    Parameters:
        results (list): A list of 'linearmodels.panel.results.PanelEffectsResults' objects containing regression results.
        options (dict, optional): A dictionary containing formatting options. Currently supports:
            - 'digits' (int): The number of decimal places to round the coefficients and t-statistics. Default is 3.
            
    Raises:
        ValueError: If 'results' is not a list of 'linearmodels.panel.results.PanelEffectsResults' objects.
        
    Returns:
        None: Prints the formatted summary table to the console.
    """

    if not are_result_type_linearmodels(results):
        raise ValueError("Only list of results of type 'linearmodels.panel.results.PanelEffectsResults' are supported.")
    
    # Extract options or use defaults
    digits = options.get('digits', 3)

    # Dependent variables
    dependent_vars = ["".join(result.model.dependent.vars)  for result in results]
    dependent_vars.insert(0, "Dependent var.")

    # Coefficients with t-stats in parentheses 
    dfs = []
    for i, result in enumerate(results):
        table = create_coefficients_table(result)[0]
        coefs = table.get("Estimate").round(digits)
        tstats = table.get("t-Statistic").round(2)
        combined = coefs.astype(str) + " (" + tstats.astype(str) + ")"
        dfs.append(combined.to_frame(name=f'Estimate_{i+1}'))

    merged_df = pd.concat(dfs, axis=1, join='outer')

    coefficients = merged_df.reset_index().fillna('').astype(str).values.tolist()

    ## Fixed effects (if any)
    included_effects = [', '.join(result.included_effects)  for result in results]
    included_effects.insert(0, "Fixed effects")

    ## VCOV type
    cov_type = [result._cov_type for result in results]
    cov_type.insert(0, "VCOV type")

    # Observations
    nobs = [f'{result.nobs:,.0f}' for result in results]
    nobs.insert(0, "Observations")

    # R2
    rsquared_inclusive = [f'{result.rsquared_inclusive:,.{digits}f}' for result in results]
    rsquared_inclusive.insert(0, "R2 (incl. FE)")

    # Within R-squared: 
    rsquared_within = [f'{result.rsquared_within:,.{digits}f}' for result in results]
    rsquared_within.insert(0, "Within R2")

    # Assemble table
    table = []
    table.append(dependent_vars)
    table.append("")
    for sublist in coefficients:
        table.append(sublist)
    table.append("")
    table.append(included_effects)
    table.append(cov_type)
    table.append(nobs)
    table.append(rsquared_inclusive)
    table.append(rsquared_within)

    output = tabulate(table, tablefmt="plain", colalign=("left",) + ("center",) * (len(results)))

    print(output)